source-code/core/src/lint/README.md

```diff
@@ -229,55 +174,100 @@ lint: {
+  visitors: {
+   Resource: ({ onLeave }) => {
++    // on leave (and many other features) could be exposed via a callback
++    onLeave(() => {
++     if (isError) report({ node: this, message: "foo" });
+    });
+   },
+  },
+ };
+});
 ```
 
-### Creating your own lint collection
+#### Use the object argument pattern to API changes
 
-A lint collection groups multiple lint rules together. You can use the provided `createLintRuleCollection` function to create a collection. By using the provided function, you will get back a strongly typed rule collection.
+To ease the iterative improvement approach, most public APIs use the object argument pattern.
 
-Example:
+```ts
+// Example: Deprecating the "level" argument.
+
+// before refactoring
+function withObjectPattern(args: { id: string; level: string; type: string })
+function withPositionalArguments(id: string, level: string, type: string)
+
+withObjectPattern({ id: "hello", level: "world", type: "JavaScript" })
+withPositionalArguments("hello", "world", "JavaScript")
+
+// after refactoring
+function withObjectPattern(args: { id: string; type: string })
+function withPositionalArguments(id: string, level?: string, type: string)
+
+withObjectPattern({ id: "hello", type: "JavaScript" })
+withPositionalArguments("hello", undefined, "JavaScript")
+```
+
+#### Users must explicitly provide provide the lint level.
+
+Letting rules define the default lint level in `createLintRule` leads to complex types, more conditional logic, and the benefit to developers is unclear. I vote to remove the "feature" due to:
+
+- Users face a hidden lint level that might make them wonder why certain things are errors or warnings.
+- Rule writers need to think about "should my rule be an error or a warning"?
 
 ```ts
-import { createLintRuleCollection } from "@inlang/core/lint"
+rules: [
+	// implicit behaviour. What rules are an error and what are a warning?
 
-const myRuleCollection = createLintRuleCollection({
-	missingKey: missingKeyRule,
-	invalidKey: invalidKeyRule,
-})
+	additionalIdRule(),
+	brandingRule(),
+	grammarRule(),
+]
+```
+
+```ts
+rules: [
+	// explicit behaviour for close to no additional "cost"
+
+	additionalIdRule("error"),
+	brandingRule("warn"),
+	grammarRule("error"),
+]
 ```
 
-The `createLintRuleCollection` expects an object where the key is the name of the provided rule and the value is the rule itself. You can add as many rules as you want.
+#### Flow of the lint module
+
+```ts
+// 1. create lint rule [Public API: Authoring rules]
+const missingMessageRule = createLintRule({ id: "inlang.missingMessage" }, ({ config, report }) => {
+	return visitors: {
+    Resource: ({ target } => {
+      report(target, { message: "Missing message" })
+    })
+	}
+})
+
+// 2. configure rule function [Public API: Using rules]
+rules: [
+  missingMessageRule("error", {
+    strict: true,
+  })
+]
+
+// 3. setup rule function [Internal]
+const rules = rules.map((rule) => await rule.setup({ config, report }))
```

Comment:
there is no `setup` function anymore

Reply:
There is. I moved `visitors` into the setup function. https://github.com/inlang/inlang/pull/459#issuecomment-1482681050